rfd/0080-piv-private-key.md

```diff
@@ -0,0 +1,392 @@
+
+```go
+// lib/client/weblogin.go
+type SSOLoginConsoleReq struct {
+  ...
+	AttestationRequest AttestationRequest `json:"attestation_request,omitempty"`
+}
+
+type CreateSSHCertReq struct {
+  ...
+	AttestationRequest AttestationRequest `json:"attestation_request,omitempty"`
+}
+
+type AuthenticateSSHUserRequest struct {
+  ...
+	AttestationRequest AttestationRequest `json:"attestation_request,omitempty"`
+}
+```
+
+For SSO login, the `AttestationRequest` field also needs to be added to each SSO auth request type (`OIDCAuthRequest`, `SAMLAuthRequest`, `GithubAuthRequest`), so we will make `AttestationRequest` a proto type.
+
+```proto
+// AttestationRequest is a request to attest an attestation request.
+message AttestationRequest {
+  oneof attestation_request {
+    // yubikey_attestation_request is a request to attest a specific YubiKey PIV slot.
+    YubiKeyAttestationRequest yubikey_attestation_request = 1;
+  }
+}
+
+// YubiKeyAttestationRequest is a request to attest a specific YubiKey PIV slot.
+// See https://pkg.go.dev/github.com/go-piv/piv-go@v1.9.0/piv#Attestation
+message YubiKeyAttestationRequest {
+  // slot_cert is an attestation certificate generated from a YubiKey PIV
+  // slot's public key and signed by the YubiKey's attestation certificate.
+  bytes slot_cert = 1;
+
+  // attestation_cert is a YubiKey's unique attestation certificate signed by a Yubico CA.
+  bytes attestation_cert = 2;
+}
+```
+
+When the Auth Server receives a login request, it will check the attached attestation request, and verify it. In addition to verifying the certificate chain for the hardware private key, the resulting attestation object will provide information about the private key, including;
+ - Device information including serial number, model, and version 
+ - Configured Touch (And PIN) Policies
+
+Currently, we are only interested in verifying the certificate chain for the public key, and checking its private key policy, so the resulting attestation response will look like:
+
+```go
+// AttestationResponse is veriried attestation data for a public key.
+type AttestationResponse struct {
+	// PublicKeyDER is the public key in PKIX, ASN.1 DER form.
+	PublicKeyDER []byte `json:"public_key"`
+	// PrivateKeyPolicy specifies the private key policy supported by the attested hardware key.
+	PrivateKeyPolicy PrivateKeyPolicy `json:"private_key_policy"`
+}
+```
+
+This `AttestationResponse` will then be checked against the user's private key policy requirement. If the policy requirement is met, the Auth server will sign the user's certificates with a private key policy extension matching the attestation. 
+
+```go
+// tls extension
+PrivateKeyPolicyASN1ExtensionOID = asn1.ObjectIdentifier{1, 3, 9999, 1, 15}
+
+// ssh extension
+CertExtensionPrivateKeyPolicy = "private-key-policy"
```

Comment:
I noticed some SSH extensions are prefixed by "teleport-" while others aren't, do you know if there's a logic to it?

Reply:
Makes sense, thanks for the explanation.

I did keep the `teleport-` in a recent design, fwiw.